File: implementation/scipamato/scipamato-core-sync/src/test/java/ch/difty/scipamato/core/sync/houskeeping/PseudoForeignKeyConstraintEnforcerTest.java
package ch.difty.scipamato.core.sync.houskeeping;

import static ch.difty.scipamato.common.TestUtils.assertDegenerateSupplierParameter;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.*;

import org.jooq.DeleteConditionStep;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.batch.core.StepContribution;
import org.springframework.batch.core.scope.context.ChunkContext;
import org.springframework.batch.repeat.RepeatStatus;

import ch.difty.scipamato.publ.db.public_.tables.records.CodeRecord;

@ExtendWith(MockitoExtension.class)
class PseudoForeignKeyConstraintEnforcerTest {

    private PseudoForeignKeyConstraintEnforcer<CodeRecord> fkce;

    @Mock
    private DeleteConditionStep<CodeRecord> stepMock;
    @Mock
    private StepContribution                contributionMock;
    @Mock
    private ChunkContext                    chunkContextMock;

    @BeforeEach
    void setUp() {
        fkce = new PseudoForeignKeyConstraintEnforcer<>(stepMock, "code", "s");
    }

    @Test
    void executing_withNullStep_doesNotThrow() {
        fkce = new PseudoForeignKeyConstraintEnforcer<>(null, "code", "s");
        assertThat(fkce.execute(contributionMock, chunkContextMock)).isEqualTo(RepeatStatus.FINISHED);
    }

    @Test
    void degenerateConstruction_withNullEntityName_throws() {
        assertDegenerateSupplierParameter(() -> new PseudoForeignKeyConstraintEnforcer<>(stepMock, null, "s"),
            "entityName");
    }

    @Test
    void executing_returnsFinishedStatus() {
        assertThat(fkce.execute(contributionMock, chunkContextMock)).isEqualTo(RepeatStatus.FINISHED);
        verify(stepMock).execute();
    }

    @Test
    void executing_withNullPlural_returnsFinishedStatus_asIfPluralWereS() {
        fkce = new PseudoForeignKeyConstraintEnforcer<>(stepMock, "code", null);
        assertThat(fkce.execute(contributionMock, chunkContextMock)).isEqualTo(RepeatStatus.FINISHED);
        verify(stepMock).execute();
    }

    @Test
    void executing_withSingleModifications_logs() {
        when(stepMock.execute()).thenReturn(1);
        fkce.execute(contributionMock, chunkContextMock);
        verify(stepMock).execute();
        // log un-asserted, log format different from the one in the next test (visual
        // assertion)
    }

    @Test
    void executing_withMultipleModifications_logs() {
        when(stepMock.execute()).thenReturn(2);
        fkce.execute(contributionMock, chunkContextMock);
        verify(stepMock).execute();
        // log un-asserted, log format different form that in the previous test (visual
        // assertion)
    }

    @Test
    void executing_withoutModifications_skipsLog() {
        when(stepMock.execute()).thenReturn(0);
        fkce.execute(contributionMock, chunkContextMock);
        verify(stepMock).execute();
        // missing log un-asserted (visual assertion)
    }

    @Test
    void executing_ignoresContributionMock() {
        fkce.execute(contributionMock, chunkContextMock);
        verifyNoMoreInteractions(contributionMock);
    }

    @Test
    void executing_ignoresChunkContextMock() {
        fkce.execute(contributionMock, chunkContextMock);
        verifyNoMoreInteractions(chunkContextMock);
    }
}

File: implementation/scipamato/core/sync/src/test/kotlin/ch/difty/scipamato/core/sync/houskeeping/PseudoForeignKeyConstraintEnforcerTest.kt
package ch.difty.scipamato.core.sync.houskeeping

import ch.difty.scipamato.publ.db.tables.records.CodeRecord
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import org.assertj.core.api.Assertions.assertThat
import org.jooq.DeleteConditionStep
import org.junit.jupiter.api.Test
import org.mockito.Mockito.verify
import org.mockito.Mockito.verifyNoMoreInteractions
import org.springframework.batch.core.StepContribution
import org.springframework.batch.core.scope.context.ChunkContext
import org.springframework.batch.repeat.RepeatStatus

internal class PseudoForeignKeyConstraintEnforcerTest {

    private val stepMock = mock<DeleteConditionStep<CodeRecord>>()
    private val contributionMock = mock<StepContribution>()
    private val chunkContextMock = mock<ChunkContext>()

    private var enforcer = PseudoForeignKeyConstraintEnforcer(stepMock, "code", "s")

    @Test
    fun executing_withNullStep_doesNotThrow() {
        enforcer = PseudoForeignKeyConstraintEnforcer(null, "code", "s")
        assertThat(enforcer.execute(contributionMock, chunkContextMock)).isEqualTo(RepeatStatus.FINISHED)
    }

    @Test
    fun executing_returnsFinishedStatus() {
        assertThat(enforcer.execute(contributionMock, chunkContextMock)).isEqualTo(RepeatStatus.FINISHED)
        verify<DeleteConditionStep<CodeRecord>>(stepMock).execute()
    }

    @Test
    fun executing_withNullPlural_returnsFinishedStatus_asIfPluralWereS() {
        enforcer = PseudoForeignKeyConstraintEnforcer(stepMock, "code", null)
        assertThat(enforcer.execute(contributionMock, chunkContextMock)).isEqualTo(RepeatStatus.FINISHED)
        verify<DeleteConditionStep<CodeRecord>>(stepMock).execute()
    }

    @Test
    fun executing_withSingleModifications_logs() {
        whenever(stepMock.execute()).thenReturn(1)
        enforcer.execute(contributionMock, chunkContextMock)
        verify(stepMock).execute()
        // log un-asserted, log format different from the one in the next test (visual assertion)
    }

    @Test
    fun executing_withMultipleModifications_logs() {
        whenever(stepMock.execute()).thenReturn(2)
        enforcer.execute(contributionMock, chunkContextMock)
        verify(stepMock).execute()
        // log un-asserted, log format different form that in the previous test (visual assertion)
    }

    @Test
    fun executing_withoutModifications_skipsLog() {
        whenever(stepMock.execute()).thenReturn(0)
        enforcer.execute(contributionMock, chunkContextMock)
        verify(stepMock).execute()
        // missing log un-asserted (visual assertion)
    }

    @Test
    fun executing_ignoresContributionMock() {
        enforcer.execute(contributionMock, chunkContextMock)
        verifyNoMoreInteractions(contributionMock)
    }

    @Test
    fun executing_ignoresChunkContextMock() {
        enforcer.execute(contributionMock, chunkContextMock)
        verifyNoMoreInteractions(chunkContextMock)
    }
}